acvm-repo/acvm/src/compiler/optimizers/witness_rebinding.rs

```diff
@@ -0,0 +1,605 @@
+use std::collections::{BTreeSet, HashMap};
+
+use crate::compiler::optimizers::GeneralOptimizer;
+use acir::{
+    circuit::{
+        brillig::{BrilligInputs, BrilligOutputs},
+        directives::Directive,
+        opcodes::{BlackBoxFuncCall, ConstantOrWitnessEnum, FunctionInput, MemOp},
+        Circuit, Opcode,
+    },
+    native_types::{Expression, Witness},
+    AcirField,
+};
+
+/// `WitnessRebindingOptimizer` will remove redundant range constraints.
+///
+/// # Example
+///
+/// Suppose we had the following circuit where witness 0 is an input and
+/// witness 2 is a return value:
+///
+/// ```
+/// EXPR [ (-1, _0) (1, _1) 0 ]
+/// EXPR [ (-1, _1) (1, _2) 0 ]
+/// ```
+///
+/// While it's not possible to determine the values of any of the witnesses, it's clear that they will all be equal.
+/// As witness 1 is internal to the circuit we can then remove it and replace any references to it with witness 0.
+///
+/// ```
+/// EXPR [ (-1, _0) (1, _2) 0 ]
+/// ```
+///
+/// This optimization pass looks for [`Opcode::AssertZero`] opcodes of this form and replaces references and replaces
+/// all references to one of these witnesses with the other to apply this equality in all other opcodes.
+pub(crate) struct WitnessRebindingOptimizer<F> {
+    circuit: Circuit<F>,
+}
+
+impl<F: AcirField> WitnessRebindingOptimizer<F> {
+    /// Creates a new `WitnessRebindingOptimizer`
+    pub(crate) fn new(circuit: Circuit<F>) -> Self {
+        Self { circuit }
+    }
+
+    fn gather_equivalent_witnesses(&self) -> HashMap<Witness, Witness> {
+        // We do not want to affect the circuit's interface so avoid optimizing away these witnesses.
+        let mut required_witnesses: BTreeSet<Witness> = self
+            .circuit
+            .private_parameters
+            .union(&self.circuit.public_parameters.0)
+            .chain(&self.circuit.return_values.0)
+            .copied()
+            .collect();
+
+        let mut equivalent_witnesses: HashMap<Witness, Witness> = HashMap::new();
+        for opcode in self.circuit.opcodes.iter().rev() {
+            #[allow(clippy::single_match)]
+            match &opcode {
+                Opcode::AssertZero(Expression { mul_terms, linear_combinations, q_c })
+                    if mul_terms.is_empty() && q_c.is_zero() && linear_combinations.len() == 2 =>
+                {
+                    let [(k1, w1), (k2, w2)]: [(F, Witness); 2] =
+                        linear_combinations.clone().try_into().unwrap();
+                    if k1 * k2 == -F::one() {
+                        match (required_witnesses.contains(&w1), required_witnesses.contains(&w2)) {
```

Comment:
1. Am I reading this right? "If only W has been encountered, remap the other witness to W, else join in encountered order."
2. Could an `A -> B -> C` equality be possible here?
  - If so, it might make sense to insert witnesses sorted